code/game/objects/items/tcg/tcg.dm

```diff
@@ -0,0 +1,273 @@
+
+/datum/card/New(list/data, list/templates = list())
+	applyTemplates(data, templates)
+	apply(data)
+
+/datum/card/proc/apply(list/data)
+	for(var/name in (vars & data))
+		vars[name] = data[name]
+
+/datum/card/proc/applyTemplates(list/data, list/templates = list())
+	apply(templates["default"])
+	apply(templates[data["template"]])
+
+/obj/item/tcgcard/attack_self(mob/user)
+	. = ..()
+	to_chat(user, "<span_class='notice'>You turn the card over.</span>")
+	if(flipped == 0)
+		name = "Trading Card"
+		desc = "It's the back of a trading card... no peeking!"
+		icon_state = "cardback"
+		flipped = 1
+	else
+		name = GLOB.card_list["[id]"].name
+		desc = GLOB.card_list["[id]"].desc
+		icon_state = GLOB.card_list["[id]"].icon_state
+		flipped = 0
+
+/obj/item/tcgcard/equipped(mob/user, slot, initial)
+	. = ..()
+	transform = matrix()
+
+/obj/item/tcgcard/dropped(mob/user, silent)
+	. = ..()
+	transform = matrix(0.3,0,0,0,0.3,0)
+
+/obj/item/cardpack
+	name = "Trading Card Pack: Coder"
+	desc = "Contains six complete fuckups by the coders. Report this on github please!"
+	icon = 'icons/obj/tcg.dmi'
+	icon_state = "cardback_nt"
+	w_class = WEIGHT_CLASS_TINY
+	var/series = "MEME" //Mirrors the card series.
+	var/contains_coin = -1 //Chance of the pack having a coin in it.
+	///The amount of cards each pack contains
+	var/card_count = 6
+	///The guaranteed rarity table, acts about the same as the rarity table. it can have as many or as few raritys as you'd like
+	var/list/guar_rarity = list(
+		"legendary" = 10,
+		"epic" = 15,
+		"rare" = 25,
+		"uncommon" = 50)
+	var/list/rarityTable = list(
+		"common" = 900,
+		"uncommon" = 300,
+		"rare" = 100,
+		"epic" = 25,
+		"legendary" = 10,
+		"misprint" = 1)//The rarity table, the set must contain at least one of each
+
+/obj/item/cardpack/series_one
+	name = "Trading Card Pack: Series 1"
+	desc = "Contains six cards of varying rarity from the 2560 Core Set. Collect them all!"
+	icon = 'icons/obj/tcg.dmi'
+	icon_state = "cardpack_series1"
+	series = "coreset2020"
+	contains_coin = 10
+
+/obj/item/cardpack/resin
+	name = "Trading Card Pack: Resin Frontier Booster Pack"
+	desc = "Contains six cards of varying rarity from the Resin Frontier set. Collect them all!"
+	icon = 'icons/obj/tcg_xenos.dmi'
+	icon_state = "cardpack_resin"
+	series = "resinfront"
+	contains_coin = 0
+	rarityTable = list(
+		"common" = 900,
+		"uncommon" = 300,
+		"rare" = 100,
+		"epic" = 25,
+		"legendary" = 10)
+
+/obj/item/cardpack/Initialize()
+	. = ..()
+	transform = matrix(0.4,0,0,0,0.4,0)
+
+/obj/item/cardpack/equipped(mob/user, slot, initial)
+	. = ..()
+	transform = matrix()
+
+/obj/item/cardpack/dropped(mob/user, silent)
+	. = ..()
+	transform = matrix(0.4,0,0,0,0.4,0)
+
+/obj/item/cardpack/attack_self(mob/user)
+	. = ..()
+	var/list/datum/card/cards = buildCardListWithRarity(card_count, guar_rarity, GLOB.card_list)
+	for(var/datum/card/template in cards)
```

Comment:
Can the list `cards` contain a type different than `/datum/card/template` ?
If not, there's no need to type-check during the `for` loop.